File: internal/cloud/portal/snap_test.go
package portal

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"

	"github.com/JoseAFlores777/ccp/internal/cloud/crypt"
	"github.com/JoseAFlores777/ccp/internal/snapshot"
)

// snapVectors: lo que el portal tiene que reproducir para poder FABRICAR un
// snapshot. El id local de un manifiesto lo recalcula la máquina al guardarlo
// (snapshot.Store.SaveManifest), así que un JSON «equivalente» no basta: el
// del navegador tiene que salir byte a byte como el de json.Marshal.
type snapVectors struct {
	AK           []byte `json:"ak"`
	ManifestJSON string `json:"manifest_json"`
	Manifest     any    `json:"manifest"`
	LocalID      string `json:"local_id"`
	CloudID      string `json:"cloud_id"`
	Parent       string `json:"parent"`
	CloudParent  string `json:"cloud_parent"`
	BlobHash     string `json:"blob_hash"`
	BlobID       string `json:"blob_id"`
	BlobPlain    []byte `json:"blob_plain"`
	BlobSealed   []byte `json:"blob_sealed"`
	SealedMan    []byte `json:"sealed_manifest"`
	SnapSig      []byte `json:"snap_sig"`
	SignPub      []byte `json:"sign_pub"`
	Revision     struct {
		ID       string `json:"id"`
		Prev     string `json:"prev"`
		Device   string `json:"device"`
		Snapshot string `json:"snapshot"`
		Base     string `json:"base"`
		Sig      []byte `json:"sig"`
	} `json:"revision"`
}

func buildSnapVectors(t *testing.T) snapVectors {
	t.Helper()
	ak := make([]byte, 32)
	for i := range ak {
		ak[i] = byte(i * 7)
	}
	acct, err := crypt.NewAccount(ak)
	if err != nil {
		t.Fatal(err)
	}
	// Contenido incómodo a propósito: `<` y `&` los escapa encoding/json y no
	// JSON.stringify, un `meta` obliga a ordenar las claves del mapa, y el
	// padre y el home ejercitan los omitempty.
	m := &snapshot.Manifest{
		Format: snapshot.FormatVersion,
		Parent: "aa" + "00000000000000000000000000000000000000000000000000000000000000",
		// UTC y con segundos enteros: es lo que el portal sabe escribir, y lo
		// que Go vuelve a leer sin cambiar ni un byte.
		Created: time.Date(2026, 9, 21, 10, 30, 0, 0, time.UTC),
		Machine: "mac de <ana> & cía", Home: "/Users/ana", CCPVersion: "2.9.0", Trigger: "portal",
		Items: []snapshot.Item{
			{LPath: "ccp/ccp.yaml", Hash: snapshot.Hash([]byte("yaml")), Size: 4, Mode: 0o644, Class: snapshot.ClassAuthored},
			{LPath: "ccp/profiles/work/overlay/CLAUDE.md", Hash: snapshot.Hash([]byte("md")), Size: 2, Mode: 0o644,
				Class: snapshot.ClassAuthored, Meta: map[string]string{"z": "último", "a": "<primero>"}},
			{LPath: "claude/settings.json", Hash: snapshot.Hash([]byte("{}")), Size: 2, Mode: 0o600, Class: snapshot.ClassSecret},
		},
	}
	st, err := snapshot.Open(t.TempDir())
	if err != nil {
		t.Fatal(err)
	}
	// SaveManifest es quien calcula el id, y es el mismo código que ejecuta la
	// máquina al recibir el snapshot: si el portal se desvía, falla allí.
	if err := st.SaveManifest(m); err != nil {
		t.Fatal(err)
	}
	js, err := json.Marshal(m)
	if err != nil {
		t.Fatal(err)
	}
	v := snapVectors{AK: ak, ManifestJSON: string(js), LocalID: m.ID, Parent: m.Parent,
		CloudID: acct.SnapshotID(m.ID), CloudParent: acct.SnapshotID(m.Parent), SignPub: acct.SignPublic()}
	if err := json.Unmarshal(js, &v.Manifest); err != nil {
		t.Fatal(err)
	}
	v.BlobPlain = []byte("un CLAUDE.md con < y & dentro\n")
	v.BlobHash = snapshot.Hash(v.BlobPlain)
	v.BlobID = acct.BlobID(v.BlobHash)
	if v.BlobSealed, err = acct.SealBlob(v.BlobID, v.BlobPlain); err != nil {
		t.Fatal(err)
	}
	if v.SealedMan, err = acct.SealManifest(v.CloudID, js); err != nil {
		t.Fatal(err)
	}
	v.SnapSig = acct.Sign(v.CloudID, v.CloudParent, v.SealedMan)
	v.Revision.ID = "bb" + "00000000000000000000000000000000000000000000000000000000000000"
	v.Revision.Device = "5f1c0e2a-0000-4000-8000-00000000abcd"
	v.Revision.Snapshot = v.CloudID
	v.Revision.Sig = acct.SignRevision(crypt.RevisionParts{ID: v.Revision.ID, Prev: v.Revision.Prev,
		Device: v.Revision.Device, Snapshot: v.Revision.Snapshot, Base: v.Revision.Base})
	return v
}

// El portal fabrica snapshots en el navegador, y la máquina los guarda
// recalculando su id. Un JSON que no salga byte a byte como el de Go es un
// snapshot que se sube bien y luego ninguna máquina puede guardar.
func TestSnapshotDelPortalIgualQueElDeGo(t *testing.T) {
	node, err := exec.LookPath("node")
	if err != nil {
		t.Skip("sin node: no se puede ejecutar el modelo del portal")
	}
	b, err := json.Marshal(buildSnapVectors(t))
	if err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(t.TempDir(), "snap.json")
	if err := os.WriteFile(path, b, 0o600); err != nil {
		t.Fatal(err)
	}
	out, err := exec.Command(node, "snap_test.mjs", path).CombinedOutput()
	t.Logf("%s", out)
	if err != nil {
		t.Fatalf("el portal no fabrica el mismo snapshot que Go: %v", err)
	}
}

File: internal/cloud/portal/snap_test.mjs
// Comprueba que el portal fabrica el MISMO snapshot que Go: el mismo JSON de
// manifiesto, el mismo id local, los mismos ids de nube y las mismas firmas.
// Los vectores los genera snap_test.go con el código que usa `ccp`.
import { readFileSync } from 'node:fs';
import * as c from './web/js/crypto.js';
import * as s from './web/js/snap.js';

const v = JSON.parse(readFileSync(process.argv[2], 'utf8'));
let fallos = 0;
const check = (nombre, ok, extra) => {
  if (!ok) { fallos++; console.error('FALLA', nombre, extra ?? ''); } else { console.log('ok  ', nombre); }
};

const ak = c.b64(v.ak);
const keys = {
  data: await c.deriveSubkey(ak, 'ccp/v1/data'),
  ids: await c.deriveSubkey(ak, 'ccp/v1/ids'),
  sign: await c.deriveSubkey(ak, 'ccp/v1/sign'),
};

// El JSON, byte a byte. Es lo que decide el id, y el id lo recalcula la máquina.
const js = s.manifestJSON(v.manifest);
check('manifestJSON igual que json.Marshal', js === v.manifest_json,
  '\n  js: ' + js + '\n  go: ' + v.manifest_json);

check('id local igual que computeID', (await s.manifestID(v.manifest)) === v.local_id);
check('id de nube del snapshot', (await s.snapshotID(keys.ids, v.local_id)) === v.cloud_id);
check('id de nube del padre', (await s.snapshotID(keys.ids, v.parent)) === v.cloud_parent);
check('un padre vacío no tiene id', (await s.snapshotID(keys.ids, '')) === '');
check('id de nube de un blob', (await s.blobID(keys.ids, v.blob_hash)) === v.blob_id);

// Abrir lo que selló Go, y que lo que sella el portal vuelva a abrirse: el
// gzip del navegador no da los mismos bytes que el de Go, pero el hash que
// cuenta es el del contenido, no el del bulto.
const plano = await s.openBlob(keys.data, v.blob_id, c.b64(v.blob_sealed), s.MaxBlobSize);
check('abre un blob sellado por Go', plano && c.eq(plano, c.b64(v.blob_plain)));
const mio = await s.sealBlob(keys.data, v.blob_id, c.b64(v.blob_plain));
const vuelta = await s.openBlob(keys.data, v.blob_id, mio, s.MaxBlobSize);
check('el blob que sella el portal se vuelve a abrir', vuelta && c.eq(vuelta, c.b64(v.blob_plain)));
check('el hash local es el del contenido', (await s.localHash(c.b64(v.blob_plain))) === v.blob_hash);

// Ed25519 es determinista: la firma del portal sobre el mismo manifiesto
// sellado tiene que ser la misma cadena de bytes que la de Go.
const sellado = c.b64(v.sealed_manifest);
const sum = c.hex(await c.sha256(sellado));
const firma = await c.ed25519Sign(keys.sign, s.signedSnapshot(v.cloud_id, v.cloud_parent, sellado, sum));
check('firma del snapshot igual que la de Go', firma && c.eq(firma, c.b64(v.snap_sig)));

const r = v.revision;
const rsig = await s.signRevision(keys.sign,
  { id: r.id, prev: r.prev, device: r.device, snapshot: r.snapshot, base: r.base }, new Uint8Array(0));
check('firma de la revisión igual que la de Go', c.eq(rsig, c.b64(r.sig)));

// Y el camino entero: del manifiesto al cuerpo de POST /v1/snapshots.
const built = await s.buildSnapshot(keys, v.manifest);
check('buildSnapshot pone el id local', built.manifest.id === v.local_id);
check('buildSnapshot nombra el snapshot de la nube', built.in.id === v.cloud_id && built.in.parent === v.cloud_parent);
const abierto = c.xopen(keys.data, c.b64(built.in.manifest), c.utf8('manifest:' + v.cloud_id));
check('el manifiesto sellado por el portal abre', abierto && new TextDecoder().decode(abierto) === v.manifest_json.replace('"id":""', `"id":"${v.local_id}"`));
check('la firma del portal verifica', (await c.ed25519Verify(c.b64(v.sign_pub),
  s.signedSnapshot(v.cloud_id, v.cloud_parent, c.b64(built.in.manifest), c.hex(await c.sha256(c.b64(built.in.manifest)))),
  c.b64(built.in.sig))) === true);

process.exit(fallos === 0 ? 0 : 1);
